polkadot/node/core/candidate-validation/src/lib.rs

```diff
@@ -253,13 +259,33 @@ async fn run<Context>(
 	ctx.spawn_blocking("pvf-validation-host", task.boxed())?;
 
 	let mut tasks = FuturesUnordered::new();
+	let mut session_index: Option<SessionIndex> = None;
+	let mut is_next_session_authority = false;
+	// PVF host won't prepare the same code hash twice, so here we just avoid extra communication
+	let mut already_prepared_code_hashes = HashSet::new();
 
 	loop {
 		loop {
 			futures::select! {
 				comm = ctx.recv().fuse() => {
 					match comm {
-						Ok(FromOrchestra::Signal(OverseerSignal::ActiveLeaves(_))) => {},
+						Ok(FromOrchestra::Signal(OverseerSignal::ActiveLeaves(update))) => {
+							let mut sender = ctx.sender().clone();
+
+							let Some(leaf) = update.activated else { continue };
+							let new_session_index = new_session_index(&mut sender, session_index, leaf.hash).await;
+							if new_session_index.is_some() {
+								session_index = new_session_index;
+								already_prepared_code_hashes.clear();
+								is_next_session_authority = check_next_session_authority(&mut sender, keystore.clone(), leaf.hash).await;
+							}
+
+							// On every active leaf check candidates and prepare PVFs our node doesn't have yet.
+							if is_next_session_authority {
```

Comment:
Doesn't this return `true` if we are authorithy in the current session, if that's the case we need to make sure we don't prepare them then, because we will create a storm of prepares that would incapacitate the node.

Reply:
This needs to be changed to account for the fact we might be an authority in the current session which is what tipically happens in practice. 

And yeah, we could end up preparing more artifacts than needed and we need to fix that.